Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,485 @@
+      credential_change_listener);
+
+  // Defer initialization until we get the current user from the
+  // credential_change_listener. This is guaranteed to be synchronously
+  // dispatched onto our worker queue, so we will be initialized before any
+  // subsequently queued work runs.
+  auto shared_self = shared_from_this();
+  worker_queue()->Enqueue([shared_self, user_promise, settings] {
+    User user = user_promise->get_future().get();
+    shared_self->InitializeInternal(user, settings);
+  });
+
+  client_initialized_ = true;
+}
+
+void FirestoreClient::InitializeInternal(const User& user,
+                                         const Settings& settings) {
+  // Do all of our initialization on our own dispatch queue.
+  worker_queue()->VerifyIsCurrentQueue();
+  LOG_DEBUG("Initializing. Current user: %s", user.uid());
+
+  // Note: The initialization work must all be synchronous (we can't dispatch
+  // more work) since external write/listen operations could get queued to run
+  // before that subsequent work completes.
+  if (settings.persistence_enabled()) {
+    Path dir = [FSTLevelDB
+        storageDirectoryForDatabaseInfo:database_info_
+                     documentsDirectory:[FSTLevelDB documentsDirectory]];
+
+    FSTSerializerBeta* remote_serializer = [[FSTSerializerBeta alloc]
+        initWithDatabaseID:database_info_.database_id()];
+    FSTLocalSerializer* serializer =
+        [[FSTLocalSerializer alloc] initWithRemoteSerializer:remote_serializer];
+    FSTLevelDB* ldb;
+    Status level_db_status = [FSTLevelDB
+        dbWithDirectory:std::move(dir)
+             serializer:serializer
+              lruParams:LruParams::WithCacheSize(settings.cache_size_bytes())
+                    ptr:&ldb];
+    if (!level_db_status.ok()) {
+      // If leveldb fails to start then just throw up our hands: the error is
+      // unrecoverable. There's nothing an end-user can do and nearly all
+      // failures indicate the developer is doing something grossly wrong so we
+      // should stop them cold in their tracks with a failure they can't ignore.
+      [NSException
+           raise:NSInternalInconsistencyException
+          format:@"Failed to open DB: %s", level_db_status.ToString().c_str()];
+    }
+    lru_delegate_ = ldb.referenceDelegate;
+    persistence_ = ldb;
+    if (settings.gc_enabled()) {
+      ScheduleLruGarbageCollection();
+    }
+  } else {
+    persistence_ = [FSTMemoryPersistence persistenceWithEagerGC];
+  }
+
+  local_store_ = [[FSTLocalStore alloc] initWithPersistence:persistence_
+                                                initialUser:user];
+
+  auto datastore = std::make_shared<Datastore>(database_info_, worker_queue(),
+                                               credentials_provider_);
+
+  auto shared_self = shared_from_this();
+  remote_store_ = absl::make_unique<RemoteStore>(
+      local_store_, std::move(datastore), worker_queue(),
+      [shared_self](OnlineState online_state) {
```

Comment:
This creates a user visible retain cycle, where `FirestoreClient` has a strong reference to RemoteStore, whose `online_state_handler` captures a strong reference to `FirestoreClient`.

This should likely capture a weak_ptr instead to prevent the cycle.

Reply:
Done.